### PR TITLE
Fix #2032 & #3714: stabilize AndroidX workers

### LINKED ISSUE #2032: LogUploadWorkerTests sometimes fail on github actions.
Issue Reference: https://github.com/oppia/oppia-android/pull/1983#issuecomment-712896420

Possible solution reference: https://developer.android.com/topic/libraries/architecture/workmanager/advanced/coroutineworker

Screenshot 2020-10-20 at 7 58 26 PM

### LINKED ISSUE #3714: PlatformParameterSyncUpWorkerTest is very flaky
PlatformParameterSyncUpWorkerTest is failing multiple runs in a row on CI making it difficult to merge PRs.

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fix #2032
Fix #3714

Stabilizes both PlatformParameterSyncUpWorker & LogUploadWorker by using a ListenableWorker instead of CoroutineWorker. In order to ensure the tests aren't flaky, we have to fully control the execution pipeline for WorkManager. We get halfway there by using a SynchronousExecutor when setting up the work manager, but CoroutineExecutor has a "hop" using Kotlin's default executor before ``doWork()`` is called. Even though we later use ``withContext`` to hop over to our own background executor, that little hop is on an unmanaged thread that we cannot synchronize against in the test (which results in flakes).

The fix is to use ListenableWorker directly and immediately kick-off a coroutine using our own dispatcher, avoiding any hops. With this setup, the test environment utilizing both SynchronousExecutor and TestCoroutineDispatchers results in a fully managed & deterministic environment. Fortunately, there should be more or less no change to the workers' production behavior since the new implementation added here is close to equivalent to CoroutineWorker (at least for ``doWorker``; there are other aspects of CoroutineWorker that differ which we are ignoring--I think this is fine).

One note on correctness: an issue with both the new implementation of the workers & CoroutineWorker is neither anticipate the  potential of the coroutine task indefinitely hanging. We really ought to introduce a ``withTimeout``, but we can't in a way that won't reintroduce a potential flake in the test until #3715 is resolved. Given this isn't a change in existing prod behavior, I suspect that it's fine.

To verify that this fix works, I ran both workers' test suites 100 times locally and saw no failures (previously, I would see about ~20 failures for PlatformParameterSyncUpWorker and ~12 for LogUploadWorker).

Finally, to prevent this again in the future I introduced a new pattern check to prohibit using CoroutineWorker. I also added one for SettableFuture for good measure since it's challenging to use correctly, though I suspect this will rarely be hit since we prefer using coroutines over ListenableFutures.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by vinitamurthi — APPROVED** [review]

**Comment by BenHenning** [comment]
Thanks @vinitamurthi.

Looks like Gradle tests are failing from a missing dep. Will push a fix momentarily; just need to verify it locally first.

**Comment by BenHenning** [comment]
I'll also kick off a 1000x run for both test suites while I wait for the CI checks to pass with the latest commit (once pushed) just to see the long-term flakiness potential of both suites (I don't typically bother with such long runs, but given how flaky WorkManager has been in the past it's probably good to check since we have some time before the PR can get merged, anyway).

**Review by Arjupta — APPROVED** [review]
Thanks for explaining this fix. Changes LGTM.

**Comment by BenHenning** [comment]
https://github.com/oppia/oppia-android/runs/3392410805?check_suite_focus=true is a CI env flake it seems.

**Comment by BenHenning** [comment]
Given the only failure is an unrelated CI flake, merging this.

**Pull request merged**
